### ISSUE TITLE
An item with the same key has already been added. Key: Mopups.PreBaked.PopupPages.Loader.LoaderPopupPage

### ISSUE THREAD

**Issue by meopoc** [LuckyDucko/Mopups#145]
Hi,

I'm intermittently getting the following exception. I think it's when 2 LoaderPopupPages are loaded in quick succession.

System.ArgumentException: An item with the same key has already been added. Key: Mopups.PreBaked.PopupPages.Loader.LoaderPopupPage
  File "AndroidMopups.cs", line 129, in void AndroidMopups.HandleAccessibility(bool showPopup, bool disableAccessibilityHandling, PopupPage popup)
    accessibilityStates.Add(popup.GetType(), views);
  File "AndroidMopups.cs", line 39, in Task AndroidMopups.AddAsync(PopupPage page)
    HandleAccessibility(true, page.DisableAndroidAccessibilityHandling, page);
  File "PopupNavigation.cs", line 73, in async Task PopupNavigation.PushAsync(PopupPage page, bool animate)+PushPage(?)
    await PopupPlatform.AddAsync(page);
  File "PreBakedMopupService.cs", line 260, in async void <LoaderAttachAndPush>d__32<LoaderPopupPage>.MoveNext()
    await MainThread.InvokeOnMainThreadAsync(() => s_popupNavigation.PushAsync(popupModal));
  File "SafeFireAndForgetExtensions.shared.cs", line 96, in async void <HandleSafeFireAndForget>d__16<Exception>.MoveNext()
    await task.ConfigureAwait(continueOnCapturedContext);
...
(7 additional frame(s) were not displayed)

Any ideas?

TIA

**Comment by robin-schroeder-merc** [comment]
Same issue.

**Comment by EmilAlipiev** [comment]
oh this issue was very commun in Rg.Popups. i remember creating another nuget just solved this issue :D

**Comment by mf-noirsoft** [comment]
@meopoc  This is because you tried to open 2 popups with the same type, in this case "Mopups.PreBaked.PopupPages.Loader.LoaderPopupPage". This used to work a while ago, but now it fails in AndroidMopups.cs:

public Task AddAsync(PopupPage page)
    {
        HandleAccessibility(true, page.DisableAndroidAccessibilityHandling, page);
...
//! important keeps reference to pages that accessibility has applied to. This is so accessibility can be removed properly when popup is removed. #https://github.com/LuckyDucko/Mopups/issues/93
    readonly Dictionary<Type, List<Android.Views.View>> accessibilityStates = new();
    void HandleAccessibility(bool showPopup, bool disableAccessibilityHandling, PopupPage popup)
    {
        ...            
            accessibilityStates.Add(popup.GetType(), views);
...

accessibilityStates is a Dictionary, and uses the popup type as its key, hence we can't add more than 1 popup of the same type.

I had to ditch mopups for those cases when I needed to display multiple popups of the same type, for example during async loading of several tasks when I want to close each popup when its task is done, not in the order they were opened. It would be nice if this was improved in a future release.

**Comment by meopoc** [comment]
@mf-noirsoft Thanks for the update. Did you manage to find a solution that looks the same as LoaderPopupPage? Of course I don;t want to use 2 visually different loading functionality in the same app? Thanks

**Comment by mf-noirsoft** [comment]
@meopoc not really, I don't use LoaderPopupPage or any other PreBaked popups, I use my own popup page. What I did was switch my loader page to be shown using the community toolkit popups, which also allows to close specific modal pages, so visually nothing changed in my project. I'm yet to decide if I will be using mopups for the other popups, or just use the community toolkit. Good luck and hope you get the results you want!

**Comment by bhavanesh2001** [comment]
```c#
 public async Task ShowMopup(PopupPage page)
   {
       if (!MopupService.Instance.PopupStack.Any( x => x.GetType() == page.GetType()))
           await MopupService.Instance.PushAsync(page);
   }
```

**Comment by AdamDiament** [comment]
Downgrading to 1.2.0 fixes this

**Comment by malikberkane** [comment]
@AdamDiament Rolling back did fix the issue. Is it still fine to use the 1.2.0 version on a net9 maui project ?

**Comment by AdamDiament** [comment]
It works for me!

**Comment by craigwi** [comment]
As a workaround, you can also page.DisableAndroidAccessibilityHandling = true.

**Comment by meopoc** [comment]
@craigwi could you expand a little? Are there any other consequences to using this setting? What does it actually do?

**Comment by craigwi** [comment]
The function HandleAccessibility.AndroidMopups uses the value of that flag and if false, does no work.  The exception is thrown lower down in the function on line 130 (in current version of the sources).

The function appears to adjust two properties on views within the Popup: ImportantForAccessibility and DescendantFocusability.  For users who depend on accessibility capabilities, not doing these adjustments may impair their experience.  I can live with that limitation as long as the app stays running.

**Comment by LuckyDucko** [comment]
This is something i had to wrangle with when i was making test examples way back.
There is some reasoning behind it, potentially due to some bad design choices, but tasks being returned could close the wrong loader if it was in the process of being created. This rapid fire situation is exactly what i noticed. Especially if animations took a moment to let it close.

I cant remember the exact reasons why, but i think the quick solution at the time was to make it so that there could only be 1 of any type of popup on the stack. This limitation could be eventually fixed, but i havent worked in Maui Mobile dev for a while.

What i would recommend, is if you have tasks that require loading that fire out back to back, is to wrap them as a single task, and push that to a loader. It is not a good UI (IMHO, not judging as i literally did the same thing) to have loaders on top of loaders rapidly popping up. 

TLDR: this is a purposeful limitation of the Mopups, not a good limitation, just on purpose.
